File: plugins/google-analytics/proj.ios/AnalyticsGoogle.m
//
//  AnalyticsGoogle.cpp
//  PluginGoogleAnalytics
//
//  Created by Duc Nguyen on 7/7/15.
//  Copyright (c) 2015 Senspark Co., Ltd. All rights reserved.
//

#import "AnalyticsGoogle.h"
#import "GAILogger.h"
#import "GAI.h"
#import "GAIDictionaryBuilder.h"
#import "GAILogger.h"
#import "GAIFields.h"
#import "GAIEcommerceFields.h"

#define OUTPUT_LOG(...)                                                        \
    if (_debug)                                                                \
        NSLog(__VA_ARGS__);

@implementation AnalyticsGoogle

#pragma mark - Analytics Interface

- (id)init {
    self = [super init];
    if (self == nil) {
        return nil;
    }

    [self setRegisteredTrackers:[NSMutableDictionary dictionary]];
    return self;
}

- (void)dealloc {
    [self setRegisteredTrackers:nil];
}

- (void)startSession:(NSString*)appKey {
    if ([self currentTracker] != nil) {
        // Start a new session with a screenView hit.
        GAIDictionaryBuilder* builder = [GAIDictionaryBuilder createScreenView];
        [builder set:@"start" forKey:kGAISessionControl];
        [[self currentTracker] set:kGAIScreenName value:appKey];
        [[self currentTracker] send:[builder build]];
    } else {
        OUTPUT_LOG(@"Start session called w/o valid tracker.");
    }
}

- (void)stopSession {
    if ([self currentTracker] != nil) {
        GAIDictionaryBuilder* builder = [GAIDictionaryBuilder createScreenView];
        [builder set:@"end" forKey:kGAISessionControl];
        [[self currentTracker] send:[builder build]];
    } else {
        OUTPUT_LOG(@"Stop session called w/o valid tracker.");
    }
}

- (void)setSessionContinueMillis:(long)millis {
    OUTPUT_LOG(@"Not supported on iOS");
}

- (void)setCaptureUncaughtException:(BOOL)isEnabled {
    [[GAI sharedInstance] setTrackUncaughtExceptions:isEnabled];
}

- (void)setDebugMode:(BOOL)isDebugMode {
    _debug = isDebugMode;
    [[GAI sharedInstance] setDryRun:isDebugMode];
}

- (void)logError:(NSString*)errorId withMsg:(NSString*)message {
    OUTPUT_LOG(@"Not supported on iOS");
}

- (void)logEvent:(NSString*)eventId {
    OUTPUT_LOG(@"Not supported on iOS");
}

- (void)logEvent:(NSString*)eventId withParam:(NSMutableDictionary*)paramMap {
    OUTPUT_LOG(@"Not supported on iOS");
}

- (void)logTimedEventBegin:(NSString*)eventId {
    OUTPUT_LOG(@"Not supported on iOS");
}

- (void)logTimedEventEnd:(NSString*)eventId {
    OUTPUT_LOG(@"Not supported on iOS");
}

- (NSString*)getSDKVersion {
    return @"3.12";
}

- (NSString*)getPluginVersion {
    return @"0.1.0";
}

#pragma mark - Google Analytics Interface

- (void)configureTracker:(NSString*)trackerId {
    if (nil == trackerId) {
        OUTPUT_LOG(@"Null tracker id at configure time.");
        return;
    }

    OUTPUT_LOG(@"Configure with trackerId: %@", trackerId);
    [self createTracker:trackerId];
}

- (id<GAITracker>)_getTrackerWithId:(NSString*)trackerId {
    return [[self registeredTrackers] objectForKey:trackerId];
}

- (void)_registerTracker:(id<GAITracker>)tracker withId:(NSString*)trackerId {
    [[self registeredTrackers] setValue:tracker forKey:trackerId];
}

- (void)_enableTracker:(id<GAITracker>)tracker {
    [self setCurrentTracker:tracker];
}

- (void)enableTracker:(NSString*)trackerId {
    if (nil == trackerId) {
        return;
    }

    id<GAITracker> tracker = [self _getTrackerWithId:trackerId];
    if (tracker == nil) {
        OUTPUT_LOG(@"Trying to enable unknown tracker: %@", trackerId);
    } else {
        OUTPUT_LOG(@"Selected tracker: %@", trackerId);
        [self _enableTracker:tracker];
    }
}

- (void)createTracker:(NSString*)trackerId {
    id<GAITracker> tracker = [self _getTrackerWithId:trackerId];
    if (tracker == nil) {
        tracker = [[GAI sharedInstance] trackerWithTrackingId:trackerId];
        [self _registerTracker:tracker withId:trackerId];
    }

    [self _enableTracker:tracker];
}

- (void)setLogLevel:(NSNumber*)logLevel {
    [[[GAI sharedInstance] logger] setLogLevel:[logLevel unsignedIntegerValue]];
}

- (void)dispatchHits {
    [[GAI sharedInstance] dispatch];
}

- (void)dispatchPeriodically:(NSNumber*)seconds {
    [[GAI sharedInstance] setDispatchInterval:[seconds intValue]];
}

- (void)stopPeriodicalDispatch {
    // Disable periodic dispatch by setting dispatch interval to a value less
    // than 1.
    [[GAI sharedInstance] setDispatchInterval:0];
}

- (void)_setParameter:(NSString*)key value:(NSString*)value {
    id<GAITracker> tracker = [self currentTracker];
    if (tracker != nil) {
        [tracker set:key value:value];
    } else {
        OUTPUT_LOG(@"Set parameter called w/o valid tracker");
    }
}

- (void)setParameter:(NSDictionary*)dict {
    NSString* param1 = [dict objectForKey:@"Param1"];
    NSString* param2 = [dict objectForKey:@"Param2"];
    [self _setParameter:param1 value:param2];
}

- (void)sendHit:(NSDictionary*)dict {
    id<GAITracker> tracker = [self currentTracker];
    if (tracker != nil) {
        [tracker send:dict];
    } else {
        OUTPUT_LOG(@"Send hit called w/o valid tracker");
    }
}

- (void)setDryRun:(NSNumber*)isDryRun {
    [self setDebugMode:[isDryRun boolValue]];
}

- (void)enableAdvertisingTracking:(NSNumber*)enabled {
    id<GAITracker> tracker = [self currentTracker];
    if (tracker != nil) {
        [tracker setAllowIDFACollection:[enabled boolValue]];
    } else {
        OUTPUT_LOG(@"Advertising called w/o valid tracker.");
    }
}

- (void)_checkDictionary:(NSDictionary*)builtDict
                    dict:(NSDictionary*)expectedDict {
    if ([builtDict isEqualToDictionary:expectedDict]) {
        // Ok.
    } else {
        NSLog(@"Dictionary unmatched: %@ vs %@", builtDict, expectedDict);
        NSAssert(NO, @"..");
    }
}

- (void)_testTrackScreenView:(NSDictionary*)dict {
    NSDictionary* expectedDict =
        [[GAIDictionaryBuilder createScreenView] build];
    [self _checkDictionary:dict dict:expectedDict];
}

- (void)_testTrackEvent:(NSDictionary*)dict {
    NSDictionary* expectedDict =
        [[GAIDictionaryBuilder createEventWithCategory:@"category"
                                                action:@"action"
                                                 label:@"label"
                                                 value:@(1)] build];
    [self _checkDictionary:dict dict:expectedDict];
}

- (void)_testTrackTiming:(NSDictionary*)dict {
    NSDictionary* expectedDict =
        [[GAIDictionaryBuilder createTimingWithCategory:@"category"
                                               interval:@(1)
                                                   name:@"variable"
                                                  label:@"label"] build];
    [self _checkDictionary:dict dict:expectedDict];
}

- (void)_testTrackException:(NSDictionary*)dict {
    NSDictionary* expectedDict =
        [[GAIDictionaryBuilder createExceptionWithDescription:@"description"
                                                    withFatal:@(YES)] build];
    [self _checkDictionary:dict dict:expectedDict];
}

- (void)_testTrackSocial:(NSDictionary*)dict {
    NSDictionary* expectedDict =
        [[GAIDictionaryBuilder createSocialWithNetwork:@"network"
                                                action:@"action"
                                                target:@"target"] build];
    [self _checkDictionary:dict dict:expectedDict];
}

- (void)_testCustomDimensionAndMetric:(NSDictionary*)dict {
    GAIDictionaryBuilder* builder = [GAIDictionaryBuilder createScreenView];
    [builder set:@"1" forKey:[GAIFields customMetricForIndex:1]];
    [builder set:@"2" forKey:[GAIFields customMetricForIndex:2]];
    [builder set:@"5.5" forKey:[GAIFields customMetricForIndex:5]];
    [builder set:@"dimension_1" forKey:[GAIFields customDimensionForIndex:1]];
    [builder set:@"dimension_2" forKey:[GAIFields customDimensionForIndex:2]];

    NSDictionary* expectedDict = [builder build];
    [self _checkDictionary:dict dict:expectedDict];
}

- (void)_testEcommerceImpression:(NSDictionary*)dict {
    GAIEcommerceProduct* product0 =
        [[[GAIEcommerceProduct alloc] init] autorelease];
    [product0 setCategory:@"category0"];
    [product0 setId:@"id0"];
    [product0 setName:@"name0"];
    [product0 setPrice:@(1.5)];

    GAIEcommerceProduct* product1 =
        [[[GAIEcommerceProduct alloc] init] autorelease];
    [product1 setCategory:@"category1"];
    [product1 setId:@"id1"];
    [product1 setName:@"name1"];
    [product1 setPrice:@(2.5)];

    GAIEcommerceProduct* product2 =
        [[[GAIEcommerceProduct alloc] init] autorelease];
    [product2 setCategory:@"category2"];
    [product2 setId:@"id2"];
    [product2 setName:@"name2"];
    [product2 setPrice:@(3.0)];

    GAIEcommerceProduct* product3 =
        [[[GAIEcommerceProduct alloc] init] autorelease];
    [product3 setCategory:@"category3"];
    [product3 setId:@"id3"];
    [product3 setName:@"name3"];
    [product3 setPrice:@(4)];

    GAIDictionaryBuilder* builder = [GAIDictionaryBuilder createScreenView];
    [builder addProductImpression:product0
                   impressionList:@"impressionList0"
                 impressionSource:nil];
    [builder addProductImpression:product1
                   impressionList:@"impressionList0"
                 impressionSource:nil];
    [builder addProductImpression:product2
                   impressionList:@"impressionList1"
                 impressionSource:nil];
    [builder addProductImpression:product3
                   impressionList:@"impressionList1"
                 impressionSource:nil];

    NSDictionary* expectedDict = [builder build];
    [self _checkDictionary:dict dict:expectedDict];
}

- (void)_testEcommerceAction:(NSDictionary*)dict {
    GAIEcommerceProduct* product0 =
        [[[GAIEcommerceProduct alloc] init] autorelease];
    [product0 setCategory:@"category0"];
    [product0 setId:@"id0"];
    [product0 setName:@"name0"];
    [product0 setPrice:@(1.5)];

    GAIEcommerceProduct* product1 =
        [[[GAIEcommerceProduct alloc] init] autorelease];
    [product1 setCategory:@"category1"];
    [product1 setId:@"id1"];
    [product1 setName:@"name1"];
    [product1 setPrice:@(2)];

    GAIEcommerceProductAction* action =
        [[[GAIEcommerceProductAction alloc] init] autorelease];
    [action setAction:kGAIPAPurchase];
    [action setProductActionList:@"actionList"];
    [action setProductListSource:@"listSource"];
    [action setTransactionId:@"transactionId"];
    [action setRevenue:@(2.0)];

    GAIDictionaryBuilder* builder = [GAIDictionaryBuilder createScreenView];
    [builder addProduct:product0];
    [builder addProduct:product1];
    [builder setProductAction:action];

    NSDictionary* expectedDict = [builder build];
    [self _checkDictionary:dict dict:expectedDict];
}

@end

File: plugins/google-analytics/proj.android_studio/libPluginGoogleAnalytics/src/main/java/org/cocos2dx/plugin/AnalyticsGoogle.java
package org.cocos2dx.plugin;

import android.content.Context;
import android.util.Log;

import com.google.android.gms.analytics.ExceptionReporter;
import com.google.android.gms.analytics.GoogleAnalytics;
import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Tracker;
import com.google.android.gms.analytics.ecommerce.Product;
import com.google.android.gms.analytics.ecommerce.ProductAction;

import org.json.JSONException;
import org.json.JSONObject;

import java.lang.Thread.UncaughtExceptionHandler;
import java.util.HashMap;
import java.util.Hashtable;
import java.util.Iterator;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

public class AnalyticsGoogle implements InterfaceAnalytics {
    protected static final String          LOG_TAG          = AnalyticsGoogle.class.getName();
    protected              Context         _context         = null;
    protected static       boolean         _debugEnabled    = false;
    protected              GoogleAnalytics _googleAnalytics = null;
    protected              Tracker         _currentTracker  = null;

    private Map<String, Tracker> _registeredTrackers = null;

    @SuppressWarnings("unused") // JNI method.
    public AnalyticsGoogle(Context context) {
        _registeredTrackers = new HashMap<>();
        _context = context;
        _googleAnalytics = GoogleAnalytics.getInstance(context);
    }

    private void logD(String msg) {
        if (_debugEnabled) {
            Log.d(LOG_TAG, msg);
        }
    }

    private void _logI(String msg) {
        Log.i(LOG_TAG, msg);
    }

    @SuppressWarnings("unused") // JNI method.
    public void configureTracker(String trackerId) {
        logD("configureTracker: id = " + trackerId);
        if (null == trackerId) {
            logD("configureTracker: null tracker id at configure time.");
            return;
        }

        createTracker(trackerId);
    }

    private Tracker _getTrackerWithId(String trackerId) {
        return _registeredTrackers.get(trackerId);
    }

    private void _registerTracker(Tracker tracker, String trackerId) {
        _registeredTrackers.put(trackerId, tracker);
    }


    private void _enableTracker(Tracker tracker) {
        _currentTracker = tracker;
    }

    @SuppressWarnings("unused") // JNI method.
    public void enableTracker(String trackerId) {
        logD("enableTracker: id = " + trackerId);
        if (null == trackerId) {
            return;
        }

        Tracker tracker = _getTrackerWithId(trackerId);
        if (null == tracker) {
            logD("Trying to enable unknown tracker id: " + trackerId);
        } else {
            logD("Selected tracker id: " + trackerId);
            _enableTracker(tracker);
        }
    }

    @SuppressWarnings("WeakerAccess")
    public void createTracker(String trackerId) {
        logD("createTracker: id = " + trackerId);
        Tracker tracker = _getTrackerWithId(trackerId);
        if (null == tracker) {
            tracker = _googleAnalytics.newTracker(trackerId);
            _registerTracker(tracker, trackerId);
        }

        _enableTracker(tracker);
    }

    @SuppressWarnings("unused") // JNI method.
    public void setLogLevel(int logLevel) {
        logD("setLogLevel: level = " + logLevel);
        _googleAnalytics.getLogger().setLogLevel(logLevel);
    }

    @SuppressWarnings("unused") // JNI method.
    public void dispatchHits() {
        _googleAnalytics.dispatchLocalHits();
    }

    @SuppressWarnings("unused") // JNI method.
    public void dispatchPeriodically(int seconds) {
        _googleAnalytics.setLocalDispatchPeriod(seconds);
    }

    @SuppressWarnings("unused") // JNI method.
    public void stopPeriodicalDispatch() {
        _googleAnalytics.setLocalDispatchPeriod(-1);
    }

    private void _setParameter(String key, String value) {
        if (_currentTracker != null) {
            _currentTracker.set(key, value);
        }
    }

    @SuppressWarnings("unused") // JNI method.
    public void setParameter(JSONObject dict) {
        logD("setParameter: dict = " + dict);
        try {
            String param1 = dict.getString("Param1");
            String param2 = dict.getString("Param2");
            _setParameter(param1, param2);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private Map<String, String> _convertToMap(JSONObject dict) throws JSONException {
        Map<String, String> map = new HashMap<>();
        Iterator<String> keys = dict.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            String value = dict.getString(key);
            map.put(key, value);
        }
        return map;
    }

    @SuppressWarnings("unused") // JNI method.
    public void sendHit(JSONObject dict) {
        logD("sendHit: dict = " + dict);
        if (_currentTracker != null) {
            try {
                Map<String, String> map = _convertToMap(dict);
                _currentTracker.send(map);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    @SuppressWarnings("WeakerAccess")
    public void setDryRun(boolean isDryRun) {
        logD("setDryRun: enabled = " + isDryRun);
        _googleAnalytics.setDryRun(isDryRun);
    }

    @SuppressWarnings("unused")
    public void enableAdvertisingTracking(boolean enabled) {
        if (null != _currentTracker) {
            _currentTracker.enableAdvertisingIdCollection(enabled);
        } else {
            Log.e(LOG_TAG, "Advertising called w/o valid tracker.");
        }
    }

    private void _checkDictionary(Map<String, String> builtDict, Map<String, String> expectedDict) {
        if (builtDict.size() != expectedDict.size()) {
            Log.wtf(LOG_TAG, String.format(Locale.getDefault(),
                "Dictionary size mismatched: expected %d found %d", expectedDict.size(),
                builtDict.size()));
            return;
        }
        Set<String> allKeys = expectedDict.keySet();
        for (String key : allKeys) {
            String expectedValue = expectedDict.get(key);
            String value = builtDict.get(key);
            if (!expectedValue.equals(value)) {
                Log.wtf(LOG_TAG, String.format(Locale.getDefault(),
                    "Element value mismatched: expected %s found %s", expectedValue, value));
            }
        }
    }

    @SuppressWarnings("unused") // JNI method.
    private void _testTrackScreenView(JSONObject json) throws JSONException {
        logD("_testTrackScreenView");
        Map<String, String> dict = _convertToMap(json);
        Map<String, String> expectedDict = new HitBuilders.ScreenViewBuilder().build();
        _checkDictionary(dict, expectedDict);
    }

    @SuppressWarnings("unused") // JNI method.
    private void _testTrackEvent(JSONObject json) throws JSONException {
        logD("_testTrackEvent");
        Map<String, String> dict = _convertToMap(json);
        Map<String, String> expectedDict = new HitBuilders.EventBuilder()
            .setCategory("category")
            .setAction("action")
            .setLabel("label")
            .setValue(1)
            .build();
        _checkDictionary(dict, expectedDict);
    }

    @SuppressWarnings("unused") // JNI method.
    private void _testTrackTiming(JSONObject json) throws JSONException {
        logD("_testTrackTiming");
        Map<String, String> dict = _convertToMap(json);
        Map<String, String> expectedDict = new HitBuilders.TimingBuilder()
            .setCategory("category")
            .setValue(1)
            .setVariable("variable")
            .setLabel("label")
            .build();
        _checkDictionary(dict, expectedDict);
    }

    @SuppressWarnings("unused") // JNI method.
    private void _testTrackException(JSONObject json) throws JSONException {
        logD("_testTrackException");
        Map<String, String> dict = _convertToMap(json);
        Map<String, String> expectedDict =
            new HitBuilders.ExceptionBuilder().setDescription("description").setFatal(true).build();
        _checkDictionary(dict, expectedDict);
    }

    @SuppressWarnings("unused") // JNI method.
    private void _testTrackSocial(JSONObject json) throws JSONException {
        logD("_testTrackSocial");
        Map<String, String> dict = _convertToMap(json);
        Map<String, String> expectedDict = new HitBuilders.SocialBuilder()
            .setNetwork("network")
            .setAction("action")
            .setTarget("target")
            .build();
        _checkDictionary(dict, expectedDict);
    }

    @SuppressWarnings("unused") // JNI method.
    private void _testCustomDimensionAndMetric(JSONObject json) throws JSONException {
        logD("_testCustomDimensionAndMetric");
        Map<String, String> dict = _convertToMap(json);
        Map<String, String> expectedDict = new HitBuilders.ScreenViewBuilder()
            .setCustomMetric(1, 1)
            .setCustomMetric(2, 2)
            .setCustomMetric(5, 5.5f)
            .setCustomDimension(1, "dimension_1")
            .setCustomDimension(2, "dimension_2")
            .build();
        _checkDictionary(dict, expectedDict);
    }

    @SuppressWarnings("unused") // JNI method.
    private void _testEcommerceImpression(JSONObject json) throws JSONException {
        logD("_testEcommerceImpression");
        Map<String, String> dict = _convertToMap(json);

        Product product0 =
            new Product().setCategory("category0").setId("id0").setName("name0").setPrice(1.5);
        Product product1 =
            new Product().setCategory("category1").setId("id1").setName("name1").setPrice(2.5);
        Product product2 =
            new Product().setCategory("category2").setId("id2").setName("name2").setPrice(3.0);
        Product product3 =
            new Product().setCategory("category3").setId("id3").setName("name3").setPrice(4);
        Map<String, String> expectedDict = new HitBuilders.ScreenViewBuilder()
            .addImpression(product0, "impressionList0")
            .addImpression(product1, "impressionList0")
            .addImpression(product2, "impressionList1")
            .addImpression(product3, "impressionList1")
            .build();

        _checkDictionary(dict, expectedDict);
    }

    @SuppressWarnings("unused") // JNI method.
    private void _testEcommerceAction(JSONObject json) throws JSONException {
        logD("_testEcommerceAction");
        Map<String, String> dict = _convertToMap(json);

        Product product0 =
            new Product().setCategory("category0").setId("id0").setName("name0").setPrice(1.5);
        Product product1 =
            new Product().setCategory("category1").setId("id1").setName("name1").setPrice(2);
        ProductAction action = new ProductAction(ProductAction.ACTION_PURCHASE)
            .setProductActionList("actionList")
            .setProductListSource("listSource")
            .setTransactionId("transactionId")
            .setTransactionRevenue(2.0);
        Map<String, String> expectedDict = new HitBuilders.ScreenViewBuilder()
            .addProduct(product0)
            .addProduct(product1)
            .setProductAction(action)
            .build();

        _checkDictionary(dict, expectedDict);
    }

    @Override
    public void startSession(String appKey) {
        if (_currentTracker != null) {
            _currentTracker.setScreenName(appKey);
            _currentTracker.send(new HitBuilders.ScreenViewBuilder().setNewSession().build());
        } else {
            Log.e(LOG_TAG, "Start session called w/o valid tracker.");
        }
    }

    @Override
    public void stopSession() {
        if (_currentTracker != null) {
            _currentTracker.send((new HitBuilders.ScreenViewBuilder().set("&sc", "end")).build());
        } else {
            Log.e(LOG_TAG, "Start session called w/o valid tracker.");
        }
    }

    @Override
    public void setSessionContinueMillis(int millis) {
        Log.i(LOG_TAG, "Not supported on Android");
    }

    @Override
    public void setCaptureUncaughtException(boolean isEnabled) {
        if (isEnabled) {
            if (null != _currentTracker) {
                UncaughtExceptionHandler myHandler = new ExceptionReporter(_currentTracker,
                    Thread.getDefaultUncaughtExceptionHandler(), _context);
                Thread.setDefaultUncaughtExceptionHandler(myHandler);
            } else {
                Log.e(LOG_TAG, "setCaptureUncaughtException called w/o valid tracker.");
            }
        } else {
            Thread.setDefaultUncaughtExceptionHandler(null);
        }
    }

    @Override
    public void setDebugMode(boolean isDebugMode) {
        _logI("setDebugMode: enabled = " + isDebugMode);
        _debugEnabled = isDebugMode;
        setDryRun(_debugEnabled);
    }

    @Override
    public void logError(String errorId, String message) {
        Log.i(LOG_TAG, "Not supported on Android");
    }

    @Override
    public void logEvent(String eventId) {
        Log.i(LOG_TAG, "Not supported on Android");
    }

    @Override
    public void logEvent(String eventId, Hashtable<String, String> paramMap) {
        Log.i(LOG_TAG, "Not supported on Android");
    }

    @Override
    public void logTimedEventBegin(String eventId) {
        Log.i(LOG_TAG, "Not supported on Android");
    }

    @Override
    public void logTimedEventEnd(String eventId) {
        Log.i(LOG_TAG, "Not supported on Android");
    }

    @Override
    public String getSDKVersion() {
        return "7.5.71";
    }

    @Override
    public String getPluginVersion() {
        return "0.9.0";
    }
}
